### ISSUE TITLE
Winrt module not importing properly

### ISSUE THREAD

**Issue by willf668** [py2exe/py2exe#118]
I'm compiling an exe for a program that uses a specific submodule of winrt. The program works when running it normally and it compiles successfully, but when I run the .exe, the process hangs for about 15 seconds, then spits out this error:

    Traceback (most recent call last):
      File "app.py", line 11, in <module>
        from winrt.windows.media.control import GlobalSystemMediaTransportControlsSessionManager as MediaManager
      File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
      File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
      File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
      File "zipextimporter.pyc", line 97, in load_module
      File "<frozen zipimport>", line 259, in load_module
      File "winrt\windows\media\__init__.pyc", line 101, in <module>
    AttributeError: 'NoneType' object has no attribute 'MediaTimeRange'

The Winrt package is included in the build, as confirmed by the traceback and also by checking in library.zip. Not entirely sure what's happening.

### Code

    import asyncio
    import json
    import os
    import threading
    import signal
    import sys
    import time
    from pypresence import Presence
    import re

from winrt.windows.media.control import GlobalSystemMediaTransportControlsSessionManager as MediaManager`

### Setup.py
    from distutils.core import setup
    import py2exe
    setup(
        console=["app.py"],
        options={"py2exe": {
            "bundle_files": 1, "compressed": True, "optimize": 2,
        }},
        version="1.0.0"
    )

**Comment by albertosottile** [comment]
Hi, thanks for reporting this. Quick question: are you able to reproduce the error with `bundle_files": 3`?

**Comment by albertosottile** [comment]
Closing this issue as no further information were provided. In general, only standard library packages are actively supported when `"bundle_files" < 3`. Hence, I would advise you to increase the value for this option.

**Comment by RvanBaarle** [comment]
I can confirm this issue. Unfortunately using `"bundle_files": 3` does not fix it. I did however attempt some debugging and it seems that the root cause is that the `__file__` attribute is missing from a module imported from a `.pyd` file. Which can be confirmed by comparing the output of the following script ran with and without py2exe:

```python3
from winrt import _winrt


print(hasattr(_winrt, '__file__'))  # outputs True if run standalone, False if ran as .exe
```

## Why does this break WinRT?

WinRT uses a slightly exotic system to put multiple modules inside a single `.pyd` file. To stay with the example above, which fails on importing `winrt.windows.media`, these are the relevant contents of `winrt/windows/media/__init__.py`:

```python3
# WARNING: Please don't edit this file. It was generated by Python/WinRT v0.9.210202.1

import typing, winrt
import enum

_ns_module = winrt._import_ns_module("Windows.Media")

...  # irrelevant other imports and definitions

MediaTimeRange = _ns_module.MediaTimeRange  # throws AttributeError: 'NoneType' object has no attribute 'MediaTimeRange'
...  # more definitions which don't get executed
```

`_import_ns_module(ns)` is defined as follows in `winrt/__init__.py`:

```python3
# WARNING: Please don't edit this file. It was generated by Python/WinRT v0.9.210202.1

from . import _winrt
_winrt.init_apartment()

def _import_ns_module(ns):
    import importlib.machinery
    import importlib.util

    try:
        module_name = "_winrt_" + ns.replace('.', '_')

        loader = importlib.machinery.ExtensionFileLoader(module_name, _winrt.__file__)  # _winrt has no attribute __file__
        spec = importlib.util.spec_from_loader(module_name, loader)
        module = importlib.util.module_from_spec(spec)
        loader.exec_module(module)
        return module
    except:  # triggering this exception catcher
        return None
```

Because `_winrt` doesn't have a `__file__` attribute the try clause fails, causing `_import_ns_module` to return `None`, resulting in `_ns_module` in the other file to be `None` as well.

## The hacky fix
Since the `.pyd` file is still around in the filesystem, I've found that it's possible to work around the issue by manually setting the `__file__` member:
```python3
from winrt import _winrt
_winrt.__file__ = '<path to project folder>/dist/winrt._winrt.pyd'
```

## The real fix?
I don't know enough about the internals of py2exe or the python import system to say how hard it would be to get that `__file__` attribute back, or if it is at all possible. It definitely doesn't sound like something high priority.

Perhaps it would be easier to get the winrt library to stop depending on the `__file__` attribute. They could also improve their error reporting, by not solving all their error with a little `except: pass`.

**Comment by albertosottile** [comment]
Hi, thanks for this careful investigation and for confirming the issue occurs also with `"bundle_files": 3`. I will look at this as soon as I can.

**Comment by albertosottile** [comment]
Fixed in 0.12.0.0
